signac/contrib/job.py

```diff
@@ -24,31 +23,173 @@
+            return
+
+        tmp_statepoint_file = self.filename + "~"
+        should_init = False
+        try:
+            # Move the statepoint to an intermediate location as a backup.
+            os.replace(self.filename, tmp_statepoint_file)
+            try:
+                new_workspace = os.path.join(job._project.workspace(), new_id)
+                os.replace(job.workspace(), new_workspace)
+            except OSError as error:
+                os.replace(tmp_statepoint_file, self.filename)  # rollback
+                if error.errno in (errno.EEXIST, errno.ENOTEMPTY, errno.EACCES):
+                    raise DestinationExistsError(new_id)
+                else:
+                    raise
+            else:
+                should_init = True
+        except OSError as error:
+            # The most likely reason we got here is because the state point
+            # file move failed due to the job not being initialized so the file
+            # doesn't exist, which is OK.
+            if error.errno != errno.ENOENT:
+                raise
+
+        # Update each job instance.
+        for job in self._jobs:
+            old_id = job._id
+            job._id = new_id
+            job._wd = None
+            job._document = None
+            job._stores = None
+            job._cwd = []
+
+        # Since all the jobs are equivalent, just grab the filename from the
+        # last one and init it. Also migrate the lock for multithreaded support.
+        old_lock_id = self._lock_id
+        self._filename = job._statepoint_filename
+        type(self)._locks[self._lock_id] = type(self)._locks.pop(old_lock_id)
+
+        if should_init:
+            job.init()
+
+        logger.info(f"Moved '{old_id}' -> '{new_id}'.")
+
+    def save(self, force=False):
+        """Trigger a save to disk.
+
+        Unlike normal JSONDict objects, this class requires the ability to save
+        on command. Moreover, this save must be conditional on whether or not a
+        file is present to allow the user to observe state points in corrupted
+        data spaces and attempt to recover.
+
+        Parameters
+        ----------
+        force : bool
+            If True, save even if the file is present on disk.
+        """
+        try:
+            # Open the file for writing only if it does not exist yet.
+            if force or not os.path.isfile(self._filename):
+                super()._save()
+        except Exception as error:
+            if not isinstance(error, OSError) or error.errno not in (
+                errno.EEXIST,
+                errno.EACCES,
+            ):
+                # Attempt to delete the file on error, to prevent corruption.
+                # OSErrors that are EEXIST or EACCES don't need to delete the file.
+                try:
+                    os.remove(self._filename)
+                except Exception:  # ignore all errors here
+                    pass
+                raise error
```

Comment:
I can't remember how the tracebacks look for `except Exception as error: raise error` versus a bare `raise` that re-raises. I _think_ we want to prefer the bare `raise` (so that the original source of the error is what shows up first in the traceback).

Reply:
Here's an example. Using `raise error` adds an extra level to the traceback, otherwise mostly identical. In both cases the final line in the traceback is the original source. I'm fine with whichever, @bdice let me know what you think:

```
>>> def f():
...     raise ValueError("In f")
...
>>> def g():
...     try:
...             f()
...     except ValueError as error:
...             raise
...
>>> def h():
...     try:
...             f()
...     except ValueError as error:
...             raise error
...
>>> g()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in g
  File "<stdin>", line 2, in f
ValueError: In f
>>> h()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 5, in h
  File "<stdin>", line 3, in h
  File "<stdin>", line 2, in f
ValueError: In f
```

Reply:
Yes, please use a bare `raise`.
```suggestion
                raise
```